bcm53xx: add support for Iptime A5004NS

djdisodo opened the pull request:
1 usb 3.0 port
1 wan port and 4 lan port

2.4GHz wifi (14e4:a8db) bcm43217 (b/g/~~n~~) (two antennas)
~~5GHz wifi (14e4:4360) bcm4360 (b/g/n/ac) (broken) (three antennas) (not supported chip)~~

Cpu: Broadcom BCM4708 (800 MHz, 2 cores)
Serial flash: 32 MiB
Ram: 256 MiB

flashing:
i failed to break vendor's firmware check

but it has uart pin on board so you don't need to make it
it uses CFE bootloader
(it lacks some commands like tftpd)
1. connect with router with cable
2. set pc's ip manually 192.168.0.2
3. run command `flash -noheader : flash0.trx` on CFE console
4. push firmware to 192.168.0.1 with tftp

adschm commented:
Please add commit message (specifications and flashing instructions) to the commit itself.

Please do it in this commit and do _not_ close and open a new one.

adschm commented on target/linux/bcm53xx/base-files/etc/board.d/02_network:
```diff
@@ -49,6 +49,10 @@ bcm53xx_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "2:lan" "3:wan" "5@eth0"
 		;;
+	iptime,a5004ns)
```
Alphabetic sorting.

adschm commented on target/linux/bcm53xx/base-files/etc/board.d/02_network:
```diff
@@ -49,6 +49,10 @@ bcm53xx_setup_interfaces()
+	iptime,a5004ns)
+		ucidef_add_switch "switch0" \
+			"0:wan" "1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1" "5@eth0"
+                ;;
```
Broken indent.

adschm commented on target/linux/bcm53xx/image/Makefile:
```diff
@@ -452,4 +452,14 @@ define Device/tplink_archer-c9-v1
 endef
 TARGET_DEVICES += tplink_archer-c9-v1
 
+define Device/iptime_a5004ns
```
Alphabetic sorting.

adschm commented on target/linux/bcm53xx/image/Makefile:
```diff
@@ -452,4 +452,14 @@ define Device/tplink_archer-c9-v1
+  DEFICE_VENDOR = ipTIME
+  DEVICE_MODEL := A5004NS
+  DEVICE_PACKAGES := $(B43) $(USB3_PACKAGES)
+  DEVICE_DTS := bcm4708-iptime-a5004ns
```
default value for DEVICE_DTS does not work?

djdisodo replied:
i tried just removing it and i got an error
`cat /home/sodo/openwrt/build_dir/target-arm_cortex-a9_musl_eabi/linux-bcm53xx_generic/image-.dtb >> /home/sodo/openwrt/build_dir/target-arm_cortex-a9_musl_eabi/linux-bcm53xx_generic/iptime_a5004ns-kernel.bin
cat: /home/sodo/openwrt/build_dir/target-arm_cortex-a9_musl_eabi/linux-bcm53xx_generic/image-.dtb: No such file or directory`
how should i change it?

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,120 @@
+--- /dev/null
++++ b/arch/arm/boot/dts/bcm4708-iptime-a5004ns.dts
+@@ -0,0 +1,87 @@
++/*
```
You should add an SPDX license identifier here.

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,120 @@
++		compatible = "gpio-leds";
++
++		power {
++			label = "bcm53xx:blue:power";
```
Remove the prefix here, i.e. just use "blue:power" etc. for LEDs.

aparcar commented on target/linux/bcm53xx/image/Makefile:
```diff
@@ -452,4 +452,14 @@ define Device/tplink_archer-c9-v1
 TARGET_DEVICES += tplink_archer-c9-v1
 
+define Device/iptime_a5004ns
+  DEFICE_VENDOR = ipTIME
```
Typo

djdisodo replied:
@aparcar i fixed the typo but i'm still having same issue

djdisodo committed:
bcm53xx: add support for Iptime A5004NS

1 usb 3.0 port
1 wan port and 4 lan port

2.4GHz wifi:
  * (14e4:a8db) bcm43217 (b/g/n)
  * 2 antennas
  * n mode not supported by driver

5GHz wifi:
  * (14e4:4360) bcm4360 (b/g/n/ac)
  * 3 antennas
  * not supported by driver

Cpu: Broadcom BCM4708 (800 MHz, 2 cores)
Serial flash: 32 MiB
Ram: 256 MiB

flashing:
i failed to break vendor's firmware check

but it has uart pin on board so you don't need to make it
it uses CFE bootloader
(it lacks some commands like tftpd)

    1. connect with router with cable
    2. set pc's ip manually 192.168.0.2
    3. run command `flash -noheader : flash0.trx` on CFE console
    4. push firmware to 192.168.0.1 with tftp

Signed-off-by: Jisoo Lee <djdisodo@gmail.com>

mans0n commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++
++/ {
++	compatible = "iptime,a5004ns", "brcm,bcm4708";
++	model = "ipTIME A5004NS (BCM4708)";
```
Is there another variant using different SoC?

rmilecki replied:
Only some early DTS have `(soc)` in `model`, later on we stopped using that. Please drop that part.

mans0n commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++	leds {
++		compatible = "gpio-leds";
++
++		power {
```
Really? The [stock FW](https://github.com/iptime-gpl/userapps_a5004ns/blob/master/fs/bcmac/clones/a5004ns/si/hw#L31) says gpio3 is for wlan2g LED.

djdisodo replied:
@mans0n i'll remove that, i guess neither power led nor wlan2g is controllable
and use cpu led as power instead

mans0n replied:
Typically for ipTIME devices, power LED is not controllable, it is directly connected to VCC.
If you rename cpu to power then it will work out of the box but it might confuse users whether it is really a power LED or a CPU LED.

mans0n commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++		cpu {
++                        label = "blue:cpu";
++                        gpios = <&chipcommon 6 GPIO_ACTIVE_HIGH>;
++                        linux,default-trigger = "timer";
```
IMO this should be dealt with by [diag.sh](https://github.com/openwrt/openwrt/blob/master/target/linux/bcm53xx/base-files/etc/diag.sh#L16), not timer trigger.

mans0n commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++		wlan5g {
++			label = "blue:wlan5g";
++			gpios = <&chipcommon 0 GPIO_ACTIVE_HIGH>;
++			linux,default-trigger = "phy1tpt";
```
I guess the trigger would not even exist if wlan driver does not support the WiFi chip.

djdisodo replied:
then will i remove it? or leave it for future

mans0n replied:
Yes, please remove this line unless you know its support is scheduled in the near future.

mans0n reviewed (commented):
Hi, a few comments below:

rmilecki commented:
@djdisodo: please kindly send your DTS to Florian for upstream inclusion in the https://github.com/Broadcom/stblinux repository.

djdisodo commented:
@rmilecki 
hopefully in weekend i will

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++                bootargs = "console=ttyS0,115200";
++        };
++
++
```
Second empty line can be dropped.

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++                        label = "blue:cpu";
++                        gpios = <&chipcommon 6 GPIO_ACTIVE_HIGH>;
++                        linux,default-trigger = "timer";
++                };
```
Broken indent for the whole block (again). Please use only tabs in DTS.

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++		};
++	};
++
++	keys: keys {
```
This is DTS, so you won't need the "keys:" DT label.

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++                        label = "Reset";
++                        linux,code = <KEY_RESTART>;
++                        gpios = <&chipcommon 11 GPIO_ACTIVE_LOW>;
++                };
```
Broken indent again for both blocks.

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <40000000>;
++
```
Remove empty line.

adschm commented on target/linux/bcm53xx/patches-5.4/340-ARM-BCM4708-Add-DT-for-Iptime-A5004NS.patch:
```diff
@@ -0,0 +1,121 @@
+From 51d1ba0fbf01a7c216e12f10fe0ff8044b012f30 Mon Sep 17 00:00:00 2001
+From: sodo <djdisodo@gmail.com>
```
This should match you Signed-off-by.

aparcar commented:
Please rebase this PR.

mans0n commented:
@djdisodo Ping.

djdisodo commented:
i've sent my patch to upstream but wasn't able to get enough feedback,

currently i'm not working on this anymore(this router is in use now)
so even if i we're able to rebase i would not test the build

[`e9672b`](https://github.com/openwrt/openwrt/commit/e9672b1a8fa4714cbc35d5964caf704b2e571f35) changed `target/linux/bcm53xx/base-files/etc/board.d/02_network ` a lot and i don't know how should i rebase this

mans0n commented:
e9672b1a8fa4 switched from swconfig to DSA, you need to change the DTS accordingly. Please refer to [other devices](https://github.com/openwrt/openwrt/blob/master/target/linux/bcm53xx/patches-5.10/304-ARM-dts-BCM5301X-Specify-switch-ports-for-remaining-.patch).

You will not be able to proceed if you can't test it runtime. Sorry to say, I'm closing this, please request me to re-open this when you are ready to push a new version.
https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr
